content/posts/2018-01-10-kotlin-null.md

```diff
@@ -0,0 +1,249 @@
+* The Kotlin argument does not work all the time and it should be the last one you have up your sleeve.
+  Not so many people actually use Kotlin at this point, but there is a huge amount of Java code.
+* Nullability annotations improve documentation and IDE support for everyone who uses the Java code,
+  so it seems like a better argument.
+* Pull Requests are always welcome!
+
+### Facade the Hell of It
+
+The worst case scenario — you do not have access to the source code, maintainers are quiet for years.
+
+Make your own abstraction using Kotlin! Believe me, almost everything is doable with enough effort
+(excluding codebases with a huge number of public methods — like RxJava).
+As a side benefit — you will make your code more testable encapsulating implementation behind an interface.
+
+Sample time!
+
+```java
+class CorporateCalculator {
+    void calculate(Argument argument) {
+        if (argument == null) throw new NullPointerException("Gotcha!");
+
+        calculateUsingCorporateBlackMagic(argument);
+    }
+}
+```
+```kotlin
+interface Calculator {
+    fun calculate(argument: Argument)
+
+    class CorporateImpl : Calculator {
+        override fun calculate(argument: Argument) {
+            CorporateCalculator().calculate(argument)
+        }
+    }
+}
+```
+
+As you see, we hide the `CorporateCalculator` behind a Kotlin-based abstraction
+which will save us from passing `null` to the `calculate` method.
+
+# Making a Swift Turn
+
+The good news is — we are not alone. Like, on the planet!
+
+The Java and Kotlin situation is kind of similar to the Objective-C and Swift one.
+Objective-C objects are essentially C pointers, which can be `NULL` or `nil` in Objective-C terms.
+In other words, these objects are implicitly nullable.
+On another hand, Swift requires all values to have nullability defined explicitly.
+
+Sounds exactly like our case, right?
+
+Apple provides Nullability Annotations for Objective-C as well — `_Nullable` + `nullable` and `_Nonnull` + `nonnull`.
+You use them — Swift fetches them using the Objective-C and Swift bridging.
+But what happens when you don’t use annotations in your Objective-C code? This is where things are getting interesting.
+
+Instead of taking the Kotlin-like path where all Java values without annotations on them
+are getting almost a blind check (pass `null` values freely as arguments and get results as non-nullable),
+Swift declares them as [_implicitly unwrapped optionals_](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/TheBasics.html).
+In other words, Swift kind of unwraps values for you,
+but you are doing it at your own risk since it is basically an equivalent of doing `nullable!!` in Kotlin.
+At the same time, it gives you a benefit of the non-warning-hostile environment.
+Just imagine if _all_ Java arguments and results would be nullable! It would be honest, but kind of gross to use.
```

Comment:
>Just imagine if all Java arguments and results would be nullable!

Hm, I don't think I get what you want to say here. Kotlin has concept of "Platform Type" which is essentially an unknown nullability state where NPE can be thrown by JVM at invocation time as it is in Java

Reply:
I’ve meant the situation when all return values would be Kotlin-nullable, i. e. if Java returns `String` Kotlin would see it as `String?`. Does it make sense?

Reply:
I should mention the Platform Type though, it would be better comparison, thanks.

Reply:
Ah I see, got confused by `nullable` used as reference name in sentence above